File: CPP/ByTier/D/1714_D_Color_with_Occurrences.cpp
// Problem: CF 1714 D - Color with Occurrences
// https://codeforces.com/contest/1714/problem/D

/*
Algorithm: Greedy + String Matching
Approach:
- For each position in the string t, find the best possible match from the given set of strings s_i.
- At each step, greedily select the string that covers the furthest position.
- If no valid string can extend the coverage, it's impossible to color the entire string.

Time Complexity: O(n * len^3)
Space Complexity: O(len + n * len)

Where:
- n: number of strings in the set
- len: length of the text t
*/

#include <string.h>
#include <iostream>

using namespace std;
int _, i, tot, now, p, q, f, l, j, k, len, n, u[200], v[200], w[20];
char s[200], t[20][20];
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cin >> _;
  while (_--) {
    cin >> (s + 1) >> n;
    len = strlen(s + 1);
    for (i = 1; i <= n; i++)
      cin >> (t[i] + 1), w[i] = strlen(t[i] + 1);
    for (tot = now = p = q = 0; now < len;) {
      for (j = 0; j <= now; j++) // Check all positions up to current coverage
        for (k = 1; k <= n; k++) // For each string
          if (j + w[k] <= len) { // Check if string fits
            f = 1;
            for (l = 1; l <= w[k]; l++) // Match characters
              if (s[j + l] != t[k][l])
                f = 0;
            if (f && j + w[k] > p + w[q]) // If match and extends furthest
              p = j, q = k;
          }
      if (p + w[q] <= now) { // If no progress made
        cout << "-1" << "\n";
        goto br; // Impossible to cover
      }
      u[++tot] = p + 1; // Record start position (1-indexed)
      v[tot] = q;       // Record string index
      now = p + w[q];   // Extend coverage
    }
    cout << tot << "\n";
    for (i = 1; i <= tot; i++)
      cout << v[i] << " " << u[i] << "\n";
  br:;
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1714/1714_D_Color_with_Occurrences.py
# Problem: CF 1714 D - Color with Occurrences
# https://codeforces.com/contest/1714/problem/D

"""
Algorithm: Greedy
Purpose: Given a text and a set of strings, color all characters of the text red by covering 
         substrings with the given strings. Find the minimum number of operations needed.
         
Time Complexity: O(|t| * n * |s|) where |t| is the length of the text, n is number of strings, 
                 and |s| is the average length of the strings.
Space Complexity: O(n * |s|) for storing the strings.

The approach uses a greedy method:
1. For each position in the text, check all strings to see if they match starting at that position.
2. Keep track of the furthest reachable position (rr) and the best string to extend coverage.
3. When we cannot extend coverage (rr == i), it's impossible to color the entire text.
4. Record the operations needed in the answer.
"""

def f():
    t = input()  # Read the text

    n = int(input())  # Read number of strings
    s = [input() for i in range(n)]  # Read all strings
    
    r = (0, -1)  # Initialize r to track the best extension (furthest reach, string index, start position)
    ans = []  # To store the operations

    rr = 0  # Current furthest reachable position
    
    for i in range(len(t)):  # Iterate through each character in the text
        for j, x in enumerate(s):  # Check each string in the set
            if t[i : i + len(x)] == x:  # If the string matches at position i
                r = max(r, (i + len(x), j, i))  # Update r to the furthest reach with this string
        if i == rr:  # When we reach the last reachable position
            if r[0] <= i:  # If we cannot extend further, it's impossible
                print(-1)
                return
            rr = r[0]  # Update furthest reach
            ans.append((r[2], r[1] + 1))  # Store the operation (start pos, string index+1)
    
    print(len(ans))  # Print number of operations
    for l, r in ans:  # Print each operation
        print(r, l + 1)  # r is string index+1, l is start position (0-based)


for _ in range(int(input())):  # Process multiple test cases
    f()  # Run function for each test case


# https://github.com/VaHiX/CodeForces/